src/dvc_data/diff.py

```diff
@@ -122,12 +126,14 @@ def _in_cache(oid, cache):
             ret.modified.append(change)
         elif change.typ == DELETE:
             ret.deleted.append(change)
+        elif (
+            consider_not_in_cache_as_modified
+            and not change.new.in_cache
+            and not (change.new.oid and change.new.oid.isdir)
+        ):
+            assert change.typ == UNCHANGED
+            ret.modified.append(change)
         else:
             assert change.typ == UNCHANGED
-            if not change.new.in_cache and not (
-                change.new.oid and change.new.oid.isdir
-            ):
-                ret.modified.append(change)
-            else:
-                ret.unchanged.append(change)
```

Comment:
IIRC we've discussed that this probably doesn't fit here and could be called manually a level above on the diff results. The ugly flag is a signal that we are doing something wrong here.

Reply:
@skshetry Make `diff` yield changes instead of packing them up into a summary. And then generate a summary for `checkout`. That will let you stril `in_cache` checks from the body of the `diff` too.

Reply:
Having a DiffResult is convenient. Also mixing up this with IO-bound checkout might not be a good thing. I think this should be evaluated separately. I could iterate on checkout over `diff.unchanged`.

Reply:
The benchmark does not show any hints of slowdown, so I am going with this simple approach instead.
```
--------------------------------------------------------------------------------- benchmark 'test_checkout[copy]': 2 tests --------------------------------------------------------------------------------
Name (time in ms)                           Min                 Max                Mean             StdDev              Median                IQR            Outliers     OPS            Rounds  Iterations
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_checkout[copy] (0001_main)        873.0456 (1.0)      986.2386 (1.04)     945.4843 (1.03)     48.8409 (1.54)     964.7036 (1.07)     77.89[35](https://github.com/iterative/dvc-data/runs/7976701078?check_suite_focus=true#step:8:36) (1.33)          1;0  1.0577 (0.97)          5           1
test_checkout[copy] (0002_0722711)     879.[38](https://github.com/iterative/dvc-data/runs/7976701078?check_suite_focus=true#step:8:39)96 (1.01)     948.9804 (1.0)      913.5191 (1.0)      31.6151 (1.0)      905.7897 (1.0)      58.4157 (1.0)           2;0  1.0947 (1.0)
```

Reply:
Or do you need those?

Reply:
I just don't remember if you need all that _in_cache info. If you don't - might be a very good opportunity to only call it on stuff that you care about.

Reply:
I think overall it makes sense, not to `cache.check` in the diff. But I am not sure how it'll affect as `TreeEntry` comparison depends on it. Will try looking into it, I think if all tests pass, should be okay to remove that.

Reply:
@efiop, let's merge this. Even `checkout` requires some work, as we want to always check for cache before we checkout, or remove file (where we prompt if the cache is missing).